File: src/main/java/com/procurement/notice/model/ocds/Lot.java
package com.procurement.notice.model.ocds;

import com.fasterxml.jackson.annotation.*;
import java.util.List;
import javax.validation.Valid;
import lombok.Getter;
import lombok.Setter;
import org.apache.commons.lang.builder.EqualsBuilder;
import org.apache.commons.lang.builder.HashCodeBuilder;

@Getter
@Setter
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonPropertyOrder({
        "id",
        "title",
        "description",
        "status",
        "statusDetails",
        "value",
        "options",
        "recurrentProcurement",
        "renewals",
        "variants",
        "contractPeriod",
        "placeOfPerformance"
})
public class Lot {
    @JsonProperty("id")
    @JsonPropertyDescription("A local identifier for this lot, such as a lot number. This is used in relatedLot " +
            "references at the item, document and award level.")
    private final String id;

    @JsonProperty("title")
    @JsonPropertyDescription("A title for this lot.")
    private final String title;

    @JsonProperty("description")
    @JsonPropertyDescription("A description of this lot.")
    private final String description;

    @JsonProperty("status")
    @JsonPropertyDescription("The current status of the process related to this lot based on the [tenderStatus " +
            "codelist](http://ocds.open-contracting.org/standard/r/1__0__0/en/schema/codelists#tender-status)")
    private TenderStatus status;

    @JsonProperty("statusDetails")
    @JsonPropertyDescription("Additional details of status.)")
    private TenderStatusDetails statusDetails;

    @JsonProperty("value")
    @Valid
    private final Value value;

    @JsonProperty("options")
    @JsonPropertyDescription("Details about lot options: if they will be accepted and what they can consist of. " +
            "Required by the EU")
    @Valid
    private final List<Option> options;

    @JsonProperty("recurrentProcurement")
    @JsonPropertyDescription("Details of possible recurrent procurements and their subsequent calls for competition.")
    @Valid
    private final List<RecurrentProcurement> recurrentProcurement;

    @JsonProperty("renewals")
    @JsonPropertyDescription("Details of allowable contract renewals")
    @Valid
    private final List<Renewal> renewals;

    @JsonProperty("variants")
    @JsonPropertyDescription("Details about lot variants: if they will be accepted and what they can consist of. " +
            "Required by the EU")
    @Valid
    private final List<Variant> variants;

    @JsonProperty("contractPeriod")
    @Valid
    private final Period contractPeriod;

    @JsonProperty("placeOfPerformance")
    @Valid
    private final PlaceOfPerformance placeOfPerformance;

    @JsonCreator
    public Lot(@JsonProperty("id") final String id,
               @JsonProperty("title") final String title,
               @JsonProperty("description") final String description,
               @JsonProperty("status") final TenderStatus status,
               @JsonProperty("statusDetails") final TenderStatusDetails statusDetails,
               @JsonProperty("value") final Value value,
               @JsonProperty("options") final List<Option> options,
               @JsonProperty("recurrentProcurement") final List<RecurrentProcurement> recurrentProcurement,
               @JsonProperty("renewals") final List<Renewal> renewals,
               @JsonProperty("variants") final List<Variant> variants,
               @JsonProperty("contractPeriod") final Period contractPeriod,
               @JsonProperty("placeOfPerformance") final PlaceOfPerformance placeOfPerformance) {
        this.id = id;
        this.title = title;
        this.description = description;
        this.status = status;
        this.statusDetails = statusDetails;
        this.value = value;
        this.options = options;
        this.recurrentProcurement = recurrentProcurement;
        this.renewals = renewals;
        this.variants = variants;
        this.contractPeriod = contractPeriod;
        this.placeOfPerformance = placeOfPerformance;
    }

    @Override
    public int hashCode() {
        return new HashCodeBuilder().append(id)
                .append(title)
                .append(description)
                .append(status)
                .append(statusDetails)
                .append(value)
                .append(options)
                .append(recurrentProcurement)
                .append(renewals)
                .append(variants)
                .append(contractPeriod)
                .append(placeOfPerformance)
                .toHashCode();
    }

    @Override
    public boolean equals(final Object other) {
        if (other == this) {
            return true;
        }
        if (!(other instanceof Lot)) {
            return false;
        }
        final Lot rhs = (Lot) other;
        return new EqualsBuilder().append(id, rhs.id)
                .append(title, rhs.title)
                .append(description, rhs.description)
                .append(status, rhs.status)
                .append(statusDetails, rhs.statusDetails)
                .append(value, rhs.value)
                .append(options, rhs.options)
                .append(recurrentProcurement, rhs.recurrentProcurement)
                .append(renewals, rhs.renewals)
                .append(variants, rhs.variants)
                .append(contractPeriod, rhs.contractPeriod)
                .append(placeOfPerformance, rhs.placeOfPerformance)
                .isEquals();
    }
}

File: src/main/kotlin/com/procurement/notice/model/ocds/Lot.kt
package com.procurement.notice.model.ocds

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty

data class Lot @JsonCreator constructor(

    val id: String,

    @JsonInclude(JsonInclude.Include.NON_NULL)
    val internalId: String?,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    val title: String?,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    val description: String?,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    var status: String?,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    var statusDetails: String?,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    val value: Value?,

    @field:JsonInclude(JsonInclude.Include.NON_EMPTY)
    @Deprecated(message = "Will not be used anymore")
    val recurrentProcurement: List<RecurrentProcurement>? = null,

    @field:JsonInclude(JsonInclude.Include.NON_EMPTY)
    @Deprecated(message = "Will not be used anymore")
    val renewals: List<Renewal>? = emptyList(),

    @field:JsonInclude(JsonInclude.Include.NON_EMPTY)
    @Deprecated(message = "Will not be used anymore")
    val variants: List<Variant>? = emptyList(),

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    val contractPeriod: Period?,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    val placeOfPerformance: PlaceOfPerformance?,

    @field:JsonInclude(JsonInclude.Include.NON_NULL)
    @field:JsonProperty("classification") @param:JsonProperty("classification") val classification: Classification?,

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @field:JsonProperty("hasOptions") @param:JsonProperty("hasOptions") val hasOptions: Boolean?,

    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @field:JsonProperty("options") @param:JsonProperty("options") val options: List<Option>?,

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @field:JsonProperty("hasRecurrence") @param:JsonProperty("hasRecurrence") val hasRecurrence: Boolean?,

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @field:JsonProperty("recurrence") @param:JsonProperty("recurrence") val recurrence: Recurrence?,

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @field:JsonProperty("hasRenewal") @param:JsonProperty("hasRenewal") val hasRenewal: Boolean?,

    @JsonInclude(JsonInclude.Include.NON_NULL)
    @field:JsonProperty("renewal") @param:JsonProperty("renewal") val renewal: RenewalV2?
)
